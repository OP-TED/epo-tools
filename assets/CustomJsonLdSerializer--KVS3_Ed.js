import{S as e}from"./index-D8tqZlAQ.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
