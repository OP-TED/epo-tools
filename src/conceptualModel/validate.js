import { hasPrefix } from '../prefix/prefix.js'
import { ATTRIBUTE, INHERITANCE } from './const.js'

function validateNode (node) {
  return hasPrefix(node.name) ? [] : [
    {
      severity: 'error', desc: `no prefix for ${node.name}`,
    }]
}

function validateEdge (edge) {
  const {
    type, source, predicate, target, quantifiers: { quantifiersDeclared },
  } = edge
  const result = []

  const requiresTargetPrefix = type !== ATTRIBUTE
  const requiresPredicate = type !== INHERITANCE
  const requiresQuantifiers = type !== INHERITANCE

  if (!quantifiersDeclared && requiresQuantifiers) {
    result.push({
      severity: 'warning', desc: `No quantifiers defined`,
    })
  }
  if (!source) {
    result.push({
      severity: 'error', desc: `No source defined`,
    })
  } else if (!hasPrefix(source)) {
    result.push({
      severity: 'error', desc: `No prefix for source [${source}]`,
    })
  }

  if (requiresPredicate) {
    if (!predicate) {
      result.push({
        severity: 'error', desc: `No predicate defined`,
      })
    } else if (!hasPrefix(predicate)) {
      result.push({
        severity: 'error', desc: `No prefix for predicate [${predicate}]`,
      })
    }
  }

  if (!target) {
    result.push({
      severity: 'error', desc: `No target defined`,
    })
  } else if (requiresTargetPrefix && !hasPrefix(target)) {
    result.push({
      severity: 'error', desc: `No prefix for target [${target}]`,
    })
  }

  return result
}

export { validateNode, validateEdge }
