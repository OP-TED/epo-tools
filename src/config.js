const EPO_LATEST = {
  owner: 'OP-TED',
  repo: 'ePO',
  localDirectory: 'assets/ePO/master',
  latest: true,
}

const EPO_3_1_0 = {
  owner: 'OP-TED',
  repo: 'ePO',
  tag: 'v3.1.0',
  localDirectory: `assets/ePO/v3.1.0`,
}

const COMMON_VOCABS = {
  localDirectory: 'assets/common-vocabularies',
}

const branchName = 'develop'
const UNDER_REVIEW = {
  owner: 'OP-TED',
  repo: 'ePO',
  localDirectory: `assets/ePO/${branchName}`,
  branch: branchName,
}

export { EPO_LATEST, EPO_3_1_0, UNDER_REVIEW, COMMON_VOCABS }
