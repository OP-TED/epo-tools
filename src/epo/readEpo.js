import { readFileSync } from 'fs'
import { bufferToJson } from '../conceptualModel/ea-to-json.js'
import { filterBy } from '../conceptualModel/filter.js'

function getJson ({ localPath, conceptualModelPath }) {
  const databasePath = `${localPath}/${conceptualModelPath}`
  const buffer = readFileSync(databasePath)
  return bufferToJson({ buffer })
}

const narrowToEpo = (g) => filterBy(g, { filter: ['epo*'] })

function getEpoJson ({ localPath, conceptualModelPath }) {
  return narrowToEpo(getJson({ localPath, conceptualModelPath }))
}

export { getJson, getEpoJson }
