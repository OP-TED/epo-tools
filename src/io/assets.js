import formats from '@rdfjs/formats'
import fs from 'fs'
import { glob } from 'glob'
import rdf from 'rdf-ext'

function getMimetype (path) {
  if (path.endsWith('.rdf')) {
    return 'application/rdf+xml'
  }
  return 'text/turtle'
}

function getGraph (path) {
  return rdf.namedNode(`file://${path}`)
}

async function applyGlob (globPattern) {
  const files = await glob(globPattern, {
    stat: true, nodir: true,
  })

  if (files.length === 0) {
    console.log('No files found')
    console.log({ workingDir: process.cwd(), globPattern })
  } else {
    console.log('loading', files.length, 'files')
  }
  return files
}

async function getRdfAssets ({ globPattern }, graphFactory = getGraph) {

  const turtleFiles = await applyGlob(globPattern)

  const assets = []
  for (const path of turtleFiles) {

    const fileStream = fs.createReadStream(path, 'utf-8')
    const dataset = rdf.dataset()
    try {
      const mimeType = getMimetype(path)
      await dataset.import(formats.parsers.import(mimeType, fileStream))
      for (const quad of [...dataset]) {
        quad.graph = graphFactory(path)
      }
    } catch (e) {
      // console.error(path, 'failed to parse')
      throw e
    }
    assets.push({ path, dataset })
  }
  return assets
}

export { getRdfAssets, applyGlob }
