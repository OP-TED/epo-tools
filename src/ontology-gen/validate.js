function addNodeWarnings (node) {
  const warnings = hasCurie(node.name) ? [] : [
    {
      severity: 'error', desc: `no prefix for ${node.name}`,
    }]
  return { ...node, warnings }
}

function addEdgeWarnings (edge) {
  const {
    source, predicate, target, isLiteral, noQuantifiers,
  } = edge
  const warnings = []
  const requiresQuantifiers = predicate !== 'rdfs:subClassOf'
  const requiresTargetPrefix = !isLiteral

  if (noQuantifiers && requiresQuantifiers) {
    warnings.push({
      severity: 'error', desc: `No quantifiers defined`,
    })
  }
  if (!source) {
    warnings.push({
      severity: 'error', desc: `No source defined`,
    })
  } else if (!hasCurie(source)) {
    warnings.push({
      severity: 'error', desc: `No prefix for source [${source}]`,
    })
  }
  if (!predicate) {
    warnings.push({
      severity: 'error', desc: `No predicate defined`,
    })
  } else if (!hasCurie(predicate)) {
    warnings.push({
      severity: 'error', desc: `No prefix for predicate [${predicate}]`,
    })
  }
  if (!target) {
    warnings.push({
      severity: 'error', desc: `No target defined`,
    })
  } else if (requiresTargetPrefix && !hasCurie(target)) {
    warnings.push({
      severity: 'error', desc: `No prefix for target [${target}]`,
    })
  }

  return { ...edge, warnings }
}

function hasCurie (name) {
  if (name.split(':').length === 2) {
    return true
  }
}

export { addEdgeWarnings, addNodeWarnings }
