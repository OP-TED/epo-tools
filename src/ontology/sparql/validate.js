import { termsFromQuery } from './extract.js'
import { anyMatch } from '../views/filter.js'

function validateAgainstGraph (graph, { queryStr }) {
  const { terms, error } = termsFromQuery({ queryStr })
  return error ? { error } : {
    terms: terms.map(x => ({
      term: x, isPresent: anyMatch(graph, { filter: [x] }),
    })),
  }
}

export { validateAgainstGraph }
