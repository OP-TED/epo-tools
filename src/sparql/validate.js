import { anyMatch, getMatcher } from '../conceptualModel/filter.js'
import { termsFromQuery } from './extractFromQuery.js'

function validateAgainstGraph (graph, { queryStr, filter }) {

  const { terms, error } = termsFromQuery({ queryStr })

  return error ? { error } : {
    terms: terms.filter(getMatcher(filter).f).map(x => ({
      term: x, isPresent: anyMatch(graph, { filter: [x] }),
    })),
  }
}

export { validateAgainstGraph }
