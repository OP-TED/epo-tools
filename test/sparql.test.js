import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { readFileSync } from 'fs'
import { describe, it } from 'mocha'
import { UNDER_REVIEW } from '../src/config.js'
import { getEpoJson } from '../src/epo/readEpo.js'
import { applyGlob } from '../src/io/assets.js'

import { termsFromQuery } from '../src/sparql/extractFromQuery.js'
import { validateAgainstGraph } from '../src/sparql/validate.js'

expect.extend({ toMatchSnapshot })

const globPattern = `test/support/queries/**`
const assets = await applyGlob(globPattern)
const epoJson = getEpoJson(UNDER_REVIEW)

describe('extract terms', async () => {

  for (const path of assets) {
    const queryStr = readFileSync(path).toString()
    it(`asset:${path}`, async function () {
      const { terms } = termsFromQuery({ queryStr })
      expect(terms).toMatchSnapshot(this)
    })
  }

  it(`failed query`, async function () {
    const { terms, error } = termsFromQuery({ queryStr: 'not a query' })
    expect(error).toMatchSnapshot(this)
  })
})

describe('validate against conceptual model', async () => {

  for (const path of assets) {
    const queryStr = readFileSync(path).toString()

    it(`asset:${path}`, async function () {
      const result = validateAgainstGraph(epoJson, { queryStr })
      expect(result).toMatchSnapshot(this)
    })
  }

})


