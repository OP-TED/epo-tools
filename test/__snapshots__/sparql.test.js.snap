// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract terms asset:test/support/queries/SMEs_by_country.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:announcesRole",
  "epo:Winner",
  "epo:playedBy",
  "epo:hasBusinessSize",
]
`;

exports[`extract terms asset:test/support/queries/distribution-cpv.rq 1`] = `
Array [
  "epo:Notice",
  "epo:hasPublicationDate",
  "epo:refersToProcedure",
  "epo:Procedure",
  "epo:hasPurpose",
  "epo:Purpose",
  "epo:hasMainClassification",
  "skos:Concept",
  "skos:prefLabel",
  "euvoc:status",
  "at:CURRENT",
]
`;

exports[`extract terms asset:test/support/queries/notices_by_buyers_country.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:announcesRole",
  "epo:Buyer",
  "epo:playedBy",
  "epo:hasLegalName",
]
`;

exports[`extract terms asset:test/support/queries/tenders_foresee_subcontracting.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:isSubmittedForLot",
  "epo:foreseesSubcontracting",
]
`;

exports[`extract terms asset:test/support/queries/titles-procedure.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:refersToProcedure",
  "epo:hasTitle",
]
`;

exports[`extract terms asset:test/support/queries/winners.rq 1`] = `
Array [
  "epo:Notice",
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:describesLot",
  "epo:hasAwardedValue",
  "epo:comprisesTenderAwardOutcome",
  "epo:TenderAwardOutcome",
  "org:Organization",
  "epo:hasLegalName",
  "epo:MonetaryValue",
  "epo:hasAmountValue",
  "epo:hasCurrency",
]
`;

exports[`extract terms failed query 1`] = `
[Error: Parse error on line 1:
not a query
^
Expecting 'EOF', 'BASE', 'IRIREF', 'PREFIX', 'PNAME_NS', 'CONSTRUCT', 'DESCRIBE', 'ASK', 'SELECT', '(', 'LOAD', 'CREATE', 'INSERTDATA', 'DELETEDATA', 'DELETEWHERE', 'DELETE', 'INSERT', 'WITH', 'a', '!', '^', 'PNAME_LN', 'CLEAR', 'DROP', 'ADD', 'MOVE', 'COPY', got 'NOT']
`;

exports[`validate against conceptual model asset:test/support/queries/SMEs_by_country.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:announcesRole",
    },
    Object {
      "isPresent": true,
      "term": "epo:Winner",
    },
    Object {
      "isPresent": true,
      "term": "epo:playedBy",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasBusinessSize",
    },
  ],
}
`;

exports[`validate against conceptual model asset:test/support/queries/distribution-cpv.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:Notice",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:refersToProcedure",
    },
    Object {
      "isPresent": true,
      "term": "epo:Procedure",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasPurpose",
    },
    Object {
      "isPresent": true,
      "term": "epo:Purpose",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasMainClassification",
    },
  ],
}
`;

exports[`validate against conceptual model asset:test/support/queries/notices_by_buyers_country.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:announcesRole",
    },
    Object {
      "isPresent": true,
      "term": "epo:Buyer",
    },
    Object {
      "isPresent": true,
      "term": "epo:playedBy",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasLegalName",
    },
  ],
}
`;

exports[`validate against conceptual model asset:test/support/queries/tenders_foresee_subcontracting.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:refersToLot",
    },
    Object {
      "isPresent": true,
      "term": "epo:isSubmittedForLot",
    },
    Object {
      "isPresent": true,
      "term": "epo:foreseesSubcontracting",
    },
  ],
}
`;

exports[`validate against conceptual model asset:test/support/queries/titles-procedure.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:refersToProcedure",
    },
    Object {
      "isPresent": false,
      "term": "epo:hasTitle",
    },
  ],
}
`;

exports[`validate against conceptual model asset:test/support/queries/winners.rq 1`] = `
Object {
  "terms": Array [
    Object {
      "isPresent": true,
      "term": "epo:Notice",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasPublicationDate",
    },
    Object {
      "isPresent": true,
      "term": "epo:refersToLot",
    },
    Object {
      "isPresent": false,
      "term": "epo:describesLot",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasAwardedValue",
    },
    Object {
      "isPresent": true,
      "term": "epo:comprisesTenderAwardOutcome",
    },
    Object {
      "isPresent": true,
      "term": "epo:TenderAwardOutcome",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasLegalName",
    },
    Object {
      "isPresent": true,
      "term": "epo:MonetaryValue",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasAmountValue",
    },
    Object {
      "isPresent": true,
      "term": "epo:hasCurrency",
    },
  ],
}
`;
